tensorstore/kvstore/s3/signature.cc

```diff
@@ -0,0 +1,255 @@
+#include <algorithm>
+#include <sstream>
+#include <set>
+
+#include "absl/log/absl_check.h"
+#include "absl/strings/ascii.h"
+#include "absl/strings/cord.h"
+#include "absl/strings/str_format.h"
+#include "absl/strings/str_split.h"
+#include "absl/strings/strip.h"
+#include "absl/time/time.h"
+#include "tensorstore/kvstore/s3/signature.h"
+#include "tensorstore/internal/path.h"
+#include "tensorstore/internal/digest/sha256.h"
+#include "tensorstore/util/result.h"
+
+#include <openssl/evp.h>
+#include <openssl/hmac.h>
+
+using ::tensorstore::internal::PercentEncodeUriPath;
+using ::tensorstore::internal::PercentEncodeUriComponent;
+using ::tensorstore::internal::SHA256Digester;
+using ::tensorstore::internal::ParseGenericUri;
+using ::tensorstore::internal::ParsedGenericUri;
+
+namespace tensorstore {
+namespace internal_storage_s3 {
+
+/// Size of HMAC (size of SHA256 digest).
+constexpr static size_t kHmacSize = 32;
+
+void ComputeHmac(std::string_view key, std::string_view message, unsigned char (&hmac)[kHmacSize]){
+    unsigned int md_len = kHmacSize;
+    // Computing HMAC should never fail.
+    ABSL_CHECK(HMAC(EVP_sha256(),
+                    reinterpret_cast<const unsigned char*>(key.data()),
+                    key.size(),
+                    reinterpret_cast<const unsigned char*>(message.data()),
+                    message.size(), hmac, &md_len) &&
+               md_len == kHmacSize);
+}
+
+void ComputeHmac(unsigned char (&key)[kHmacSize], std::string_view message, unsigned char (&hmac)[kHmacSize]){
+    unsigned int md_len = kHmacSize;
+    // Computing HMAC should never fail.
+    ABSL_CHECK(HMAC(EVP_sha256(), key, kHmacSize,
+                    reinterpret_cast<const unsigned char*>(message.data()),
+                    message.size(), hmac, &md_len) &&
+               md_len == kHmacSize);
+}
+
+
+/// https://docs.aws.amazon.com/AmazonS3/latest/API/sig-v4-header-based-auth.html
+
+Result<std::string> CanonicalRequest(
```

Comment:
Since we will always be building the request, rather than receiving S3 URLs from an external source, I think it might be simpler to define an S3RequestBuilder class, that provides a similar interface to `HttpRequestBuilder` in `internal/http/http_request.h`.  The `BuildRequest` method would take some extra parameters, like the access key, region, etc., would add the authorization header, and return an `HttpRequest`.  Internally it would just store the headers as an `std::vector<std::string>` and the query parameters in an `std::vector<std::pair<std::string, std::string>>`.  `BuildRequest` would take care of sorting those vectors before computing the signature.  We can skip the lowercasing, and instead just require the caller to specify the headers and query params in lowercase to begin with.

Reply:
Thanks for pointing this out. I've written my understanding of what's required in a comment below, which I will update as we iterate. Feel free to edit or suggest what needs fleshing out. Also, please let me know if you have a preferred github workflow for a large piece of work like this.

Reply:
> Internally it would just store the headers as an `std::vector<std::string>` and the query parameters in an `std::vector<std::pair<std::string, std::string>>`. `BuildRequest` would take care of sorting those vectors before computing the signature. We can skip the lowercasing, and instead just require the caller to specify the headers and query params in lowercase to begin with.

Have taken this approach.